src/der.rs

```diff
@@ -104,9 +117,116 @@ pub(crate) fn expect_tag<'a>(
+        }
+        LONG_FORM_LEN_TWO_BYTES => {
+            let length_byte_one = usize::from(input.read_byte()?);
+            let length_byte_two = usize::from(input.read_byte()?);
+            let combined = (length_byte_one << 8) | length_byte_two;
+            if combined < LONG_FORM_LEN_ONE_BYTE_MAX {
+                return Err(Error::BadDer); // Not the canonical encoding.
+            }
+            combined
+        }
+        LONG_FORM_LEN_THREE_BYTES => {
+            let length_byte_one = usize::from(input.read_byte()?);
+            let length_byte_two = usize::from(input.read_byte()?);
+            let length_byte_three = usize::from(input.read_byte()?);
+            let combined = (length_byte_one << 16) | (length_byte_two << 8) | length_byte_three;
+            if combined < LONG_FORM_LEN_TWO_BYTES_MAX {
+                return Err(Error::BadDer); // Not the canonical encoding.
+            }
+            combined
+        }
+        LONG_FORM_LEN_FOUR_BYTES => {
+            let length_byte_one = usize::from(input.read_byte()?);
+            let length_byte_two = usize::from(input.read_byte()?);
+            let length_byte_three = usize::from(input.read_byte()?);
+            let length_byte_four = usize::from(input.read_byte()?);
+            let combined = (length_byte_one << 24)
+                | (length_byte_two << 16)
+                | (length_byte_three << 8)
+                | length_byte_four;
+            if combined < LONG_FORM_LEN_THREE_BYTES_MAX {
+                return Err(Error::BadDer); // Not the canonical encoding.
+            }
+            combined
+        }
+        _ => {
+            return Err(Error::BadDer); // We don't support longer lengths.
+        }
+    };
+
+    if length >= size_limit {
+        return Err(Error::BadDer); // The length is larger than the caller accepts.
+    }
+
+    let inner = input.read_bytes(length)?;
+    Ok((tag, inner))
 }
 
+// Long-form DER encoded lengths of two bytes can express lengths up to the following limit.
+//
+// The upstream ring::io::der::read_tag_and_get_value() function limits itself to up to two byte
+// long-form DER lengths, and so this limit represents the maximum length that was possible to
+// read before the introduction of the read_tag_and_get_value_limited function.
+pub(crate) const TWO_BYTE_DER_SIZE: usize = LONG_FORM_LEN_TWO_BYTES_MAX;
+
+// DER Tag identifiers have two forms:
+// * Low tag number form (for tags values in the range [0..30]
+// * High tag number form (for tag values in the range [31..]
+// We only support low tag number form.
+const HIGH_TAG_RANGE_START: u8 = 31;
+
+// DER length octets have two forms:
+// * Short form: 1 octet supporting lengths between 0 and 127.
+// * Long definite form: 2 to 127 octets, number of octets encoded into first octet.
+const SHORT_FORM_LEN_MAX: u8 = 128;
+
+// Leading octet for long definite form DER length expressed in second byte.
+const LONG_FORM_LEN_ONE_BYTE: u8 = 129;
```

Comment:
I think perhaps hex here would be more intuitive? Maybe it's just me.

(ie, `0x81` expresses more to me that the top bit is set and the second nibble is the count of length bytes.)

Reply:
> I think perhaps hex here would be more intuitive? Maybe it's just me.

I think that's good intuition. They were hex originally so it makes sense to me to preserve that. Fixed up.